hyperspy/drawing/widget.py

```diff
@@ -0,0 +1,896 @@
+        """Create the matplotlib patch(es), and store it in self.patch
+        """
+        if hasattr(super(WidgetBase, self), '_set_patch'):
+            super(WidgetBase, self)._set_patch()
+        # Must be provided by the subclass
+
+    def _add_patch_to(self, ax):
+        """Create and add the matplotlib patches to 'ax'
+        """
+        self._set_patch()
+        for p in self.patch:
+            ax.add_artist(p)
+            p.set_animated(hasattr(ax, 'hspy_fig'))
+        if hasattr(super(WidgetBase, self), '_add_patch_to'):
+            super(WidgetBase, self)._add_patch_to(ax)
+
+    def set_mpl_ax(self, ax):
+        """Set the matplotlib Axes that the widget will draw to. If the widget
+        on state is True, it will also add the patch to the Axes, and connect
+        to its default events.
+        """
+        if ax is self.ax:
+            return  # Do nothing
+        # Disconnect from previous axes if set
+        if self.ax is not None and self.is_on():
+            self.disconnect(self.ax)
+        self.ax = ax
+        canvas = ax.figure.canvas
+        if self.is_on() is True:
+            self._add_patch_to(ax)
+            self.connect(ax)
+            canvas.draw()
+
+    def connect(self, ax):
+        """Connect to the matplotlib Axes' events.
+        """
+        on_figure_window_close(ax.figure, self.close)
+        if self._navigating:
+            self.connect_navigate()
+
+    def connect_navigate(self):
+        """Connect to the axes_manager such that changes in the widget or in
+        the axes_manager are reflected in the other.
+        """
+        if self._navigating:
+            self.disconnect_navigate()
+        self.axes_manager.events.indices_changed.connect(self._on_navigate)
+        self._on_navigate(self.axes_manager)    # Update our position
+        self._navigating = True
+
+    def disconnect_navigate(self):
+        """Disconnect a previous naivgation connection.
+        """
+        self.axes_manager.events.indices_changed.disconnect(self._on_navigate)
+        self._navigating = False
+
+    def _on_navigate(self, axes_manager):
+        """Callback for axes_manager's change notification.
+        """
+        pass    # Implement in subclass!
+
+    def disconnect(self, ax):
+        """Disconnect from all events (both matplotlib and navigation).
+        """
+        for cid in self.cids:
+            try:
+                ax.figure.canvas.mpl_disconnect(cid)
+            except:
+                pass
+        if self._navigating:
+            self.disconnect_navigate()
+
+    def close(self, window=None):
+        """Set the on state to off (removes patch and disconnects), and trigger
+        events.closed.
+        """
+        self.set_on(False)
+        self.events.closed.trigger(self)
+
+    def draw_patch(self, *args):
+        """Update the patch drawing.
+        """
+        if hasattr(self.ax, 'hspy_fig'):
+            self.ax.hspy_fig._draw_animated()
```

Comment:
When you say "close a widget", what do you mean? That you call `widget.close()`?